packages/hub/schema/data-source.js

```diff
@@ -65,8 +74,39 @@ module.exports = class DataSource {
         this._staticModels = [];
       }
     }
+    let schemaModels = this._staticModels.filter(i => this._schemaContentTypes.includes(i.type)).map(i => `${i.type}/${i.id}`);
+    if (schemaModels.length) {
+    /*
+      TODO: We need to grandfather in the current static-model ability to specify schema if we want to limit this PR to
+      reasoning about being able to specify the new card schema only--otherwise we'll need to also implement the new
+      card indexing at the same time as well. The next step here is to remove this grandfathering so that everything uses
+      the new schema feature and as a result, everything is forced to use the new card boundaries. This can be done byr
+      uncommenting the line below and unskipping the unit test that goes along with this.
+
+      throw new Error(`The datasource '${this.id}' defines static-models that includes schema. Schema is not allowed in static-models. Found schema models: ${JSON.stringify(schemaModels)}`);
+    */
+    }
     return this._staticModels;
   }
+  get staticSchemaModels() {
```

Comment:
This seems problematic. We don't want schema.js to be a per-data-source feature. It is unlike static-model.js in that crucial way.

We need to load the schema for a card any time the card is part of our dependency graph. We don't want to need to configure the card as a data-source first.

Our loading of the `schemas` feature should work similar to how we load `middlewares` or `field-types`. Which is unlike how we load `writers` or `static-models`.

Reply:
Ok, so I think this means that in `SourcesUpdate.add()` I should look at the document that is being added to the index, and if it is a document that depends on a card for which we have a schema.js file, that is when we load the schema (before the document being added to the index is actually added). And as part of that I should be smart about loading the schema such that I don't reload a card's schema that I've loaded previously (like we do when loading the middleware stack). Is this the correct approach?

Reply:
oops, I was rereading this, when you said "dependency graph" you mean the package dependency graph (I was originally thinking that you meant we'd load do some kind of JIT card schema loading when an indexer encountered a card instance it didn't know about). So i think this is pretty straight forward: just move the logic that we use to load the schema feature models out of the `RunningIndexers._activateDataSource()`, since as you say, we should load the schema regardless of whether the data source is activated.

Reply:
(this also answered another question I had in a comment questioning why we need to activate data sources in order to use card schemas)